packages/nimble/inst/CppCode/nimOptim.cpp

```diff
@@ -0,0 +1,51 @@
+#include <R_ext/Applic.h>
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <nimble/nimOptim.h>
+
+// This wraps a NimObjectiveFn as an R `optimfn` type
+//   typedef double optimfn(int, double *, void *);
+// defined in
+//   https://svn.r-project.org/R/trunk/src/include/R_ext/Applic.h
+static double optimfn_wrapper(int n, double* par, void* fn) {
+    NimArr<1, double> nim_par;
+    nim_par.setSize(n);
+    std::copy(par, par + n, nim_par.getPtr());
+    return ((NimObjectiveFn*)fn)(nim_par);
+}
+
+// This attempts to match the behavior of optim() defined in the documentation
+//   https://stat.ethz.ch/R-manual/R-devel/library/stats/html/optim.html
+// and in the reference implementation
+//   https://svn.r-project.org/R/trunk/src/library/stats/R/optim.R
+//   https://svn.r-project.org/R/trunk/src/library/stats/src/optim.c
+//   https://svn.r-project.org/R/trunk/src/include/R_ext/Applic.h
+nimSmartPtr<OptimResultNimbleList> nimOptim(NimArr<1, double>& par,
+                                            NimObjectiveFn fn) {
+    const int n = par.dimSize(0);
+
+    nimSmartPtr<OptimResultNimbleList> result = new OptimResultNimbleList;
+    result->par = par;
+    result->counts.setSize(2);
+    result->hessian.setSize(n, n);
+
+    // Use Nelder-Mead by default.
+    double* Bvec = par.getPtr();
```

Comment:
This is not completely safe.  If `parMatrix` is n-x-n and the user wants to provide the 3rd row as the initial parameters to optim, they can give `par = parMatrix[3,]`.  NIMBLE would turn this into a NimArr that has map=TRUE.  This treats its contents as a vector of length n but accesses the contents using a stride of n.  When we need to ensure a NimArr argument coming into a C++ function uses contiguous memory (e.g. to pass it along to nmmin), we have used a `nimArrCopyIfNeeded` step.  Example here:
https://github.com/nimble-dev/nimble/blob/devel/packages/nimble/inst/CppCode/nimDists.cpp#L42